File: Problems/Week_2/3.29_큐(10845)/임용식/10845.cs
using System;
using System.Text;

namespace AlgoCSharp
{
    class Queue
    {
        private int[] container;
        private int front;
        private int rear;

        public Queue()
        {
            container = new int[10000];
            front = rear = 0;
        }

        public Queue(int size)
        {
            container = new int[size];
            front = rear = 0;
        }

        public void Push(int data)
        {
            container[rear++] = data;
        }

        public int Pop()
        {
            if (front == rear)
            {
                return -1;
            }
            return container[front++];
        }

        public int Size()
        {
            return rear - front;
        }

        public bool Empty()
        {
            return Size() == 0 ? true : false;
        }

        public int Front()
        {
            if (Empty())
            {
                return -1;
            }
            return container[front];
        }

        public int Back()
        {
            if (Empty())
            {
                return -1;
            }
            return container[rear - 1];
        }
    }

    class Program
    {
        static void Main(string[] args)
        {
            StringBuilder sb = new StringBuilder();
            int n = Convert.ToInt32(Console.ReadLine());
            Queue q = new Queue(n);

            while (n-- > 0)
            {
                string[] command = Console.ReadLine().Split();

                switch (command[0])
                {
                    case "push":
                        q.Push(Convert.ToInt32(command[1]));
                        break;
                    case "pop":
                        sb.Append(q.Pop()+"\n");
                        break;
                    case "size":
                        sb.Append(q.Size() + "\n");
                        break;
                    case "empty":
                        sb.Append(Convert.ToInt32(q.Empty()) + "\n");
                        break;
                    case "front":
                        sb.Append(q.Front() + "\n");
                        break;
                    case "back":
                        sb.Append(q.Back() + "\n");
                        break;
                }
            }
            Console.WriteLine(sb);
        }
    }
}

File: Problems/Week_2/3.29_큐(10845)/임용식/10845.c
#pragma warning(disable:4996)
#include <stdio.h>
#include <string.h>

#define TRUE 1
#define FALSE 0

typedef int		BOOL;
typedef struct _queue
{
	int container[10000];
	int front;
	int rear;
} Queue;

void	Push(Queue* q, const int data);
int		Pop(Queue* q);
int		Size(const Queue* q);
BOOL	Empty(const Queue* q);
int		Front(const Queue* q);
int		Back(const Queue* q);

int main()
{
	Queue q = { {0},0,0 };
	int n;
	scanf("%d", &n);

	while (n--)
	{
		char command[6] = "";
		scanf("%s", command);
		
		if (!strcmp(command, "push"))
		{
			int data = 0;
			scanf("%d", &data);
			Push(&q, data);
		}
		else if (!strcmp(command, "pop"))
		{
			printf("%d\n", Pop(&q));
		}
		else if (!strcmp(command, "size"))
		{
			printf("%d\n", Size(&q));
		}
		else if (!strcmp(command, "empty"))
		{
			printf("%d\n", Empty(&q));
		}
		else if (!strcmp(command, "front"))
		{
			printf("%d\n", Front(&q));
		}
		else if (!strcmp(command, "back"))
		{
			printf("%d\n", Back(&q));
		}
	}


	return 0;
}

void Push(Queue* q, const int data)
{
	q->container[q->rear++] = data;
}

int Pop(Queue* q)
{
	if (q->front == q->rear)
	{
		return -1;
	}
	return q->container[q->front++];
}

int Size(const Queue* q)
{
	return q->rear - q->front;
}

BOOL Empty(const Queue* q)
{
	return Size(q) ? FALSE : TRUE;
}

int Front(const Queue* q)
{
	if (Empty(q))
	{
		return -1;
	}
	return q->container[q->front];
}
int Back(const Queue* q)
{
	if (Empty(q))
	{
		return -1;
	}
	return q->container[q->rear - 1];
}
